### PR TITLE
RATIS-1521. NettyServerStreamRpc use EpollEventLoopGroup

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

NettyServerStreamRpc use EpollEventLoopGroup

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1521

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
@szetszwo  What do you think of this change?

File: ratis-netty/src/main/java/org/apache/ratis/netty/NettyConfigKeys.java [hunk 1]
```diff
@@ -106,6 +106,44 @@ static TimeDuration clientReplyQueueGracePeriod(RaftProperties properties) {
...
+      return getInt(properties::getInt, SERVER_WORKER_GROUP_THREADS_KEY,
+          SERVER_WORKER_GROUP_THREADS_DEFAULT, getDefaultLog(),
+          requireMin(0), requireMax(65536));
+    }
+
+    static void setServerWorkerGroupThreads(RaftProperties properties, int num) {
+      setInt(properties::setInt, SERVER_WORKER_GROUP_THREADS_KEY, num);
+    }
+
+    String SERVER_EPOLL_ENABLE_KEY = PREFIX + ".epoll.enable";
```
Comment (szetszwo) [inline ratis-netty/src/main/java/org/apache/ratis/netty/NettyConfigKeys.java]: Let's call it "use-epoll".
```
      String USE_EPOLL_KEY = PREFIX + ".use-epoll";
```

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java [hunk 2]
```diff
@@ -30,6 +30,7 @@
 import org.apache.ratis.server.DataStreamServerRpc;
 import org.apache.ratis.server.RaftServer;
 import org.apache.ratis.server.RaftServerConfigKeys;
+import org.apache.ratis.thirdparty.com.google.common.util.concurrent.ThreadFactoryBuilder;
```
Comment (szetszwo) [inline ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java:33]: Use ConcurrentUtils.newThreadFactory(..) instead.  We should avoid external dependency.

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java [hunk 3]
```diff
@@ -127,12 +131,33 @@ public NettyServerStreamRpc(RaftServer server) {
     }
 
     final int port = NettyConfigKeys.DataStream.port(properties);
+
+    final int accepterThreads = NettyConfigKeys.DataStream.serverBossGroupThreads(properties);
+    final int workerThreads = NettyConfigKeys.DataStream.serverWorkerGroupThreads(properties);
+
+    boolean epollEnable = NettyConfigKeys.DataStream.serverEpollEnable(properties) && Epoll.isAvailable();
```
Comment (szetszwo) [inline ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java:138]: Print a warning if use-epoll but !Epoll.isAvailable().

**Review by szetszwo — COMMENTED** [review]
@guohao-rosicky , this is a good idea. 
- Let's move the new conf properties to a new interface so that the server conf and the client conf are not mixing together.
- Let's also create a utility method for creating EventLoopGroup.
Some other comments inlined.  See also https://issues.apache.org/jira/secure/attachment/13039972/599_review.patch

**Comment by guohao-rosicky** [comment]
@szetszwo Has been modified, Please take a look.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by szetszwo** [comment]
@guohao-rosicky , one of the tests failed with TimeoutIOException: Timeout 3000ms; see https://github.com/apache/ratis/runs/5194575007?check_suite_focus=true#step:5:543

Could you fix it first?
```
Error:  Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 94.996 s <<< FAILURE! - in org.apache.ratis.datastream.TestNettyDataStreamStarTopologyWithGrpcCluster
Error:  testMultipleStreamsMultipleServersStepDownLeader(org.apache.ratis.datastream.TestNettyDataStreamStarTopologyWithGrpcCluster)  Time elapsed: 5.388 s  <<< ERROR!
java.util.concurrent.CompletionException: org.apache.ratis.protocol.exceptions.TimeoutIOException:
 Timeout 3000ms: Failed to send DataStreamWindowRequest:seqNum=12,DataStreamRequestHeader:clientId=client-DC7E0CEB7606,type=STREAM_DATA,id=368,offset=7387696,length=0
```

**Comment by guohao-rosicky** [comment]
> @guohao-rosicky , one of the tests failed with TimeoutIOException: Timeout 3000ms; see https://github.com/apache/ratis/runs/5194575007?check_suite_focus=true#step:5:543
> 
> Could you fix it first?
> 
> ```
> Error:  Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 94.996 s <<< FAILURE! - in org.apache.ratis.datastream.TestNettyDataStreamStarTopologyWithGrpcCluster
> Error:  testMultipleStreamsMultipleServersStepDownLeader(org.apache.ratis.datastream.TestNettyDataStreamStarTopologyWithGrpcCluster)  Time elapsed: 5.388 s  <<< ERROR!
> java.util.concurrent.CompletionException: org.apache.ratis.protocol.exceptions.TimeoutIOException:
>  Timeout 3000ms: Failed to send DataStreamWindowRequest:seqNum=12,DataStreamRequestHeader:clientId=client-DC7E0CEB7606,type=STREAM_DATA,id=368,offset=7387696,length=0
> ```

I've been trying to fix Timeout for good, but I haven't done it yet

**Comment by guohao-rosicky** [comment]
@szetszwo 

> @guohao-rosicky , one of the tests failed with TimeoutIOException: Timeout 3000ms; see https://github.com/apache/ratis/runs/5194575007?check_suite_focus=true#step:5:543
> 
> Could you fix it first?
> 
> ```
> Error:  Tests run: 4, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 94.996 s <<< FAILURE! - in org.apache.ratis.datastream.TestNettyDataStreamStarTopologyWithGrpcCluster
> Error:  testMultipleStreamsMultipleServersStepDownLeader(org.apache.ratis.datastream.TestNettyDataStreamStarTopologyWithGrpcCluster)  Time elapsed: 5.388 s  <<< ERROR!
> java.util.concurrent.CompletionException: org.apache.ratis.protocol.exceptions.TimeoutIOException:
>  Timeout 3000ms: Failed to send DataStreamWindowRequest:seqNum=12,DataStreamRequestHeader:clientId=client-DC7E0CEB7606,type=STREAM_DATA,id=368,offset=7387696,length=0
> ```

This has nothing to do with this PR, can we merge this PR first? I'm going to keep tracking the timeout problem

**Comment by szetszwo** [comment]
@guohao-rosicky , let's re-run the jobs and try to get a successful build.

**Pull request merged**